src/docfx/restore/CredentialHandler.cs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CredentialHandler
+    {
+        private const int RetryCount = 3;
+
+        private readonly Func<HttpRequestMessage, bool, Task<HttpConfig?>>[] _credentialProviders;
+        private readonly ConcurrentDictionary<string, Task<Dictionary<string, string>>> _credentialCache = new();
+
+        public CredentialHandler(params Func<HttpRequestMessage, bool, Task<HttpConfig?>>[] credentialProviders) => _credentialProviders = credentialProviders;
```

Comment:
consider break the `HttpRequestMessage` in the signature into: `Func<string, HttpConfig?, bool, Task<HttpConfig?>>` because:
- The interface is now symmetric, both input and output speaks `HttpConfig`
- The comparison in `LspCredentialProvider` becomes a reference equals check
- Can eliminate the need of `CopyHttpRequestMessage` because you know what old headers to remove